cpp/include/raft/core/interruptible.hpp

```diff
@@ -201,23 +201,35 @@ class interruptible {
   template <bool Claim>
   static auto get_token_impl(std::thread::id thread_id) -> std::shared_ptr<interruptible>
   {
-    std::lock_guard<std::mutex> guard_get(mutex_);
-    // the following constructs an empty shared_ptr if the key does not exist.
-    auto& weak_store  = registry_[thread_id];
+    // Make a local copy of the shared pointer to make sure the registry is not destroyed,
+    // if, for any reason, this function is called at program exit.
+    std::shared_ptr<registry_t> shared_registry = registry_;
+    // If the registry is not available, create a lone token that cannot be accessed from
+    // the outside of the thread.
+    if (!shared_registry) { return std::shared_ptr<interruptible>{new interruptible()}; }
+    // Otherwise, proceed with the normal logic
+    std::lock_guard<std::mutex> guard_get(std::get<0>(*shared_registry));
+    // the following two lines construct an empty shared_ptr if the key does not exist.
+    auto& weak_store  = std::get<1>(*shared_registry)[thread_id];
     auto thread_store = weak_store.lock();
     if (!thread_store || (Claim && thread_store->claimed_)) {
+      std::weak_ptr<registry_t> weak_registry = shared_registry;
       // Create a new thread_store in two cases:
       //  1. It does not exist in the map yet
       //  2. The previous store in the map has not yet been deleted
-      thread_store.reset(new interruptible(), [thread_id](auto ts) {
-        std::lock_guard<std::mutex> guard_erase(mutex_);
-        auto found = registry_.find(thread_id);
-        if (found != registry_.end()) {
-          auto stored = found->second.lock();
-          // thread_store is not moveable, thus retains its original location.
-          // Not equal pointers below imply the new store has been already placed
-          // in the registry_ by the same std::thread::id
-          if (!stored || stored.get() == ts) { registry_.erase(found); }
+      thread_store.reset(new interruptible(), [weak_registry, thread_id](auto ts) {
```

Comment:
Would a custom deleter be useful here, maybe? I'm wondering if maybe this lambda could be instead turned into a functor that culd have its own state and deleter to clean things up. Would that make it easier to manage the valid state here?

Reply:
Good point! I have doubts this could help with our issue, but perhaps it could make the code a bit more readable.